### ISSUE TITLE
[HOLD for payment 2023-03-01] [$1000] Trailing line break is not working inside three backticks messages on the web

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14694]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open app & navigate to any chat
2. Send a message that starts and ends with three backticks and includes a line break before closing 3 backticks

## Expected Result:
line break should reflect in the message

## Actual Result:
line break is being ignored and there is no space at the end of the message ends. 

## NOTE
What appears to be happening is that one less line break is shows than added so... 
1 line break added doesn't show
2 line breaks added shows 1
and so on, so the deliverable is to fix all instances of "one less line break added" and **not** just the instance of a single line break not showing

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] MacOS / Chrome / Safari

**Version Number:** 1.2.62-1
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
![Simulator Screen Shot - iPhone 13 - 2023-01-31 at 01 57 54](https://user-images.githubusercontent.com/43996225/215819288-44fdb8af-db2a-465f-88da-3e336ea8ec87.png)
<img width="1300" alt="Screenshot 2023-01-31 at 1 57 33 AM" src="https://user-images.githubusercontent.com/43996225/215819307-70ca2e6f-1a01-40b1-9126-deb5a510ad92.png">

https://user-images.githubusercontent.com/43996225/215819338-8c01b40e-83cb-45d4-9b7a-51a0f729f813.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @jayeshmangwani 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675110851241419

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01888fcd2426613ce9</li>
        <li>Upwork Job ID: 1620562001590956032</li>
        <li>Last Price Increase: 2023-01-31</li>
    </ul>
</details>

**Comment by mallenexpensify** [comment]
Tested, can confirm it's happening.  I updated the OP

> ## NOTE
> What appears to be happening is that one less line break is shows than added so... 
> 1 line break added doesn't show
> 2 line breaks added shows 1
> and so on, so the deliverable is to fix all instances of "one less line break added" and **not** just the instance of a single line break not showing

@jayeshmangwani you'll still be compensated as the bug reporter if we fix this, just wanted to make sure all instances were addressed.  

Moving to engineer to confirm this can be `External`, 90% sure it can...

**Comment by eh2077** [comment]
**Proposal**

RCA
Code fence
```

test

```
converts to html
```html
<pre><br />test<br /></pre>
```
But html will not display the last newline represented by the last `<br />` tag.

**Solution**
Convert code fence 
```

test

```
to html
```html
<pre><br />test<br /><br /></pre>
```

If you want to verify the solution, you can apply the following diff to expensify common module
```diff
diff --git a/lib/ExpensiMark.js b/lib/ExpensiMark.js
index 7f94ffe..b6cb064 100644
--- a/lib/ExpensiMark.js
+++ b/lib/ExpensiMark.js
@@ -31,7 +31,7 @@ export default class ExpensiMark {
                 // &nbsp; will create styling issues so use &#32;
                 replacement: (match, __, textWithinFences) => {
                     const group = textWithinFences.replace(/(?:(?![\n\r])\s)/g, '&#32;');
-                    return `<pre>${group}</pre>`;
+                    return `<pre>${group}<br /></pre>`;
                 },
             },
 
@@ -268,7 +268,7 @@ export default class ExpensiMark {
             {
                 name: 'codeFence',
                 regex: /<(pre)(?:"[^"]*"|'[^']*'|[^'">])*>([\s\S]*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))/gi,
-                replacement: '```\n$2\n```',
+                replacement: '```\n$2```',
             },
             {
                 name: 'anchor',
```

**Comment by mananjadhav** [comment]
@marktoman's proposal looks cleaner and simple to implement. I think we should go with that.

@bondydaa what do you think?

**Comment by bondydaa** [comment]
Yeah I think we'd prefer to fix this with the regex. that one is a doosy

so I would also suggest making sure we update the tests https://github.com/Expensify/expensify-common/blob/aaf49708420187a13ebe3bdb97b67104e6a6bcfc/__tests__/ExpensiMark-Markdown-test.js to also verify the extra line(s) being persevered.

**Comment by eh2077** [comment]
Hi @mananjadhav and @bondydaa 

I think my proposal will be the better solution because it avoids breaking following regression cases
1. Send following code fence and edit the sent message. There will be no extra newline in the editing markdown text


````html
```
test
```
````
2. Send following code fence and edit the send message. There will be only one trailing new line in the editing markdown text.
````html
```
test

```
````
I also recorded a video to show it

https://user-images.githubusercontent.com/117511920/216200197-97d9d82d-46ab-4095-b2ff-04adb091dc8d.mp4

As the video shows, my proposal will ensure the line break of _**both the html and markdown**_ are correct. 

But @marktoman 's proposal will bring regression for markdown line break issue.

Really hope you would take the above into consideration.

Thanks for the reviewing!

**Comment by marktoman** [comment]
@eh2077 That is another unrelated bug, which is when you edit a code block, it adds a new line. Your solution strips every last line whether it should be there or not, so while it appears to fix both problems, in reality it introduces regression for every case where the last line is supposed to be there.

**Comment by marktoman** [comment]
I've added a solution for the other bug.

**Comment by eh2077** [comment]
> @eh2077 That is another unrelated bug, which is when you edit a code block, it adds a new line. Your solution strips every last line whether it should be there or not, so while it appears to fix both problems, in reality it introduces regression for every case where the last line is supposed to be there.

The extra line break `<br />` I add will ensure to translate markdown with right trailing line break from html. So my proposal will not bring the regression you mentioned.

I'd like to compare the differences of our proposal through following cases based on the latest updated version of your proposal.

### Case 1
Currently by sending a new message, below code fence
````html
```
test

```
````
is translated to 
```html
<pre><br />test<br /></pre>
```
and both our solution will translate the markdown into 
```html
<pre><br />test<br /><br /></pre>
```
Note the `html` will be saved by the backend and for frontend UI rendering. So both of our proposals will save new structure data to backend for new messages.

### Case 2
Currently if you edit a sent message, there'll be an additional trailing line break in the markdown. As you pointed out, this is an another unrelated bug. Both of our proposals will fix it for new saved messages and historical messages. 

### Case 3
Currently if you edit a sent message and save changes(no matter has changes or not), no extra trailing line break will be added to `html` to be saved. Based on your latest updated, both of our proposals will not break it.

### Case 4
For historical saved code fence messages, both of our proposal can't get enough information to fix the potential previous missing trailing line break. If you want to test it, you can send the following two code fence block before applying the patch
````html
```
original code fence with two trailing line breaks, which will show one trailing line break


```
````
````html
```
original code fence with three trailing line breaks, which will show two trailing line breaks



```
````

### Conclusion 
Though both of the proposals fix the issue, my proposal has the following advantages 

**1. Simple. It's easy to understand and doesn’t add extra complexity to the magic regex expression 
2. More efficient. It'll be more efficient without adding extra complex regex matching**

**Comment by marktoman** [comment]
@eh2077 No, the expression (including the old one) has the new line as optional. You always add it whether it is there or not, which introduces a regression. I pass it through only if it is there. ` ```test``` ` translates to `<pre>test</pre>`, which can be e.g., copied as HTML to the clipboard via the context menu. Given each rule has separate use cases, it has to work properly on its own and not only as a pair for the specific use case of this issue. (The clipboard can be inspected here: https://evercoder.github.io/clipboard-inspector/)

In the HTML to MD conversion, if the content was added differently than through MD, e.g., imported or some future source, you don't make sure the ending triple tick occupies a new line, which introduces a regression because the original code does that.

There is nothing magical about regular expressions. You could say that very complex expressions are harder to read, but adding `(\n?)` to fix things properly for once and all hardly qualifies. And you can test the efficiency if you really believe that. The measurable difference in string ops like this is non-existent.

My solution fixes the causes rather than the effects, which makes it clean and future-proof.

**Comment by mananjadhav** [comment]
Thanks for the discussion folks. @marktoman You're assigned this job. Feel free to create the PR.

**Comment by marktoman** [comment]
@eh2077 Thanks for noticing the other bug.

**Comment by marktoman** [comment]
The PR is ready (not sure if you would know, given it's another repo)
cc @mananjadhav @bondydaa

**Comment by bondydaa** [comment]
no sweat, in the future I'd say feel free to just add the c+ and internal employee that were assigned on the App issue to the Expensify-Common PR.

**Comment by bondydaa** [comment]
Created a discussion here https://expensify.slack.com/archives/C049HHMV9SM/p1675377618968189?thread_ts=1675110851.241419&cid=C049HHMV9SM to follow up and make sure preserving the new lines is always what we want.

I agree with the comments in the PR and that we should do it the way the PR has it coded but just want to make sure I didn't miss some context from a while ago. 

I think if we don't hear anything on it by tomorrow we can merge and deal with the consequences (ie: new "bug" reports) if they come up.

**Comment by mallenexpensify** [comment]
Lotta action in here. I should hire @marktoman in Upwork, right?

**Comment by bondydaa** [comment]
doesn't look like any comments yet, tagged jason and tom on the thread to double check but I'm going to take Mark's PR off hold now.

**Comment by bondydaa** [comment]
okay https://github.com/Expensify/expensify-common/pull/496 is merged.

@marktoman you'll need to create another PR in App so that we start using the latest version of expensify-common, please add me and @mananjadhav as the reviewer (the automation might do that already?)

**Comment by bondydaa** [comment]
I forgot we had to do that, probably didn't need to hold the expensify-common PR - in the future feel free to remind engineers of this if we hold PRs needlessly.

**Comment by MelvinBot** [comment]
@mananjadhav, @bondydaa, @mallenexpensify, @marktoman Huh... This is 4 days overdue. Who can take care of this?

**Comment by MelvinBot** [comment]
@mananjadhav, @bondydaa, @mallenexpensify, @marktoman Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by bondydaa** [comment]
🤫 melvin, just merged the PR a little bit ago.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.75-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/14822


If no regressions arise, payment will be issued on 2023-03-01. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
📣 @MelvinBot! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@mananjadhav / @bondydaa] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@mananjadhav / @bondydaa] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@mananjadhav / @bondydaa] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@marktoman] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@mallenexpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by marktoman** [comment]
Regression Test Proposal

1. Add a new comment:
````
```
text

```
````
2. Verify that that the rendered box has an empty line below "text".
3. Edit the comment and verify that the text is the same as 1.
4. Repeat 1-3 for the below, but instead, make sure that it has no added line below "text":
````
```
text
```
````

Do we agree 👍 or 👎

**Comment by MelvinBot** [comment]
📣 @marktoman! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by marktoman** [comment]
Contributor details
Your Expensify account email: marktomandev@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01ebe3709b787c23aa

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by mallenexpensify** [comment]
@marktoman @jayeshmangwani @mananjadhav can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01888fcd2426613ce9

**Comment by mananjadhav** [comment]
Accepted @mallenexpensify

**Comment by mananjadhav** [comment]
@bondydaa I was looking at the original PR that caused this issue and it looks like the first newline handling was done in https://github.com/Expensify/expensify-common/commit/36531c018c1a269607a536922b107d810a026268 (PR https://github.com/Expensify/expensify-common/pull/376). But I wouldn't mark it is a regression PR because the regex did what was expected as per the GH issue.


Let me know what you think.

**Comment by bondydaa** [comment]
👍 yeah I agree with that assessment.

**Comment by marktoman** [comment]
I left an Upwork-specific question on the contract proposal @mallenexpensify

**Comment by mallenexpensify** [comment]
Replied there @marktoman

**Comment by marktoman** [comment]
Accepted @mallenexpensify

**Comment by mallenexpensify** [comment]
Thanks for the test steps @marktoman 

@mananjadhav do you have read-only access to TestRail yet? If so, can you access this - https://expensify.testrail.io/index.php?/cases/view/1971213 ?
If so, do you agree with the proposed addition below?

1. Update title from "Markup" to "Markdown" 
2. Add new section "Code Block"
3. Add a new comment:

````
```
text

```
````

4. **Verify** that that the rendered box has an empty line below "text".
5. Edit the comment and verify that the text is the same as 1.
6. Repeat 1-3 for the below, but instead, make sure that it has no added line below "text":
````
```
text
```
````

**Comment by mallenexpensify** [comment]
@marktoman it appears you were hired on 2/1 and the [PR was merged](https://github.com/Expensify/App/pull/14822#event-8565009270) on 2/20 which is 13 biz days. We have
> Merged PR more than 9 business days after assignment - 50% penalty

Before I subtract 50% from the pay for you and @mananjadhav , do either of you have reasoning to provide why it shouldn't be deducted?   Thanks

**Comment by marktoman** [comment]
@mallenexpensify There were two PRs: 1) to expensify-common, 2) to App. The first one merged quickly and covered everything discussed here. On the second PR, I found a separate issue. I would expect the second problem to occur as a new issue at some point if I would not solve it under this one.

**Comment by mananjadhav** [comment]
@mallenexpensify As @marktoman mentioned, there were two PRs and here's link of the [problem](https://github.com/Expensify/App/pull/14822#issuecomment-1416931893) they reported in rendering, which took sometime to [resolve](https://github.com/Expensify/App/pull/14822#issuecomment-1433529747). Plus I was out sick during that period that I wasn't helpful in resolve it faster.

**Comment by mallenexpensify** [comment]
Thanks for the context.  @mananjadhav what do you propose the payment breakdown should be here?

**Comment by mananjadhav** [comment]
Quick question @mallenexpensify, if we would've put the issue on Hold when we found the issue, would it still impact the payment here? If the answer is that it wouldn't have impacted, then would be great if we can make an exception here and do the 1000$ payout. If the answer is, it would still impact during the hold period, then I think it's fair to have the 50% deduction penalty.

**Comment by mallenexpensify** [comment]
Thanks @mananjadhav , this appears to be an edge case because of the non-E/App repo and because the PR was put on hold which slowed things down.   I propose we do the full $1000 payout for @marktoman and @mananjadhav for C+.  @bondydaa and/or @mananjadhav can you add a comment to confirm you're OK with this then I'll issue payments.

**Comment by mananjadhav** [comment]
I am definitely okay ;) Should we wait for @bondydaa to confirm?

**Comment by MelvinBot** [comment]
@mananjadhav, @bondydaa, @mallenexpensify, @marktoman Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@mananjadhav, @bondydaa, @mallenexpensify, @marktoman Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by mallenexpensify** [comment]
Thanks @mananjadhav , I didn't piece 2-2 together when I saw that Slack post last week.  I _get_ the reasoning behind 'repos we control', I'm unsure we make it clear in CONTRIBUTING.md or other guides how we manage multiple PRs (nor.. if we have to, at least right now).   The one part I'm unsure on is that the two PRs weren't able to be worked on in tandem, correct? If that's so, then the contributor shouldn't be _penalized_ while they're waiting for one PR to be merged or hit production because they're able to post the next PR.  @bondydaa can you take a peek when you're back on tomorrow?

**Comment by mananjadhav** [comment]
I think the difference here is we fixed the reported issue, and also a probable future issue and that held the PR. The more we discuss, I feel we shouldn't penalize for solving it right. Still waiting for @bondydaa to chime in on this one.

**Comment by mananjadhav** [comment]
@mallenexpensify @bondydaa do we have any update here? Been pending since long.

**Comment by bondydaa** [comment]
Sorry was OOO most of the past 2 weeks. 

Yep I don't think we should reduce the pay here, any delays here were due to the 2 PRs or people being OOO so not really in the control of those involved so I think $1000 makes sense.

**Comment by mallenexpensify** [comment]
Thanks @bondydaa and everyone else for the patience 

@jayeshmangwani paid $250 for reporting
@mananjadhav paid $1000 for C+
@marktoman paid $1000 for contributor 

📕

**Comment by mananjadhav** [comment]
Thanks @mallenexpensify and @bondydaa

**Comment by marktoman** [comment]
@mallenexpensify 
> The one part I'm unsure on is that the two PRs weren't able to be worked on in tandem, correct? If that's so, then the contributor shouldn't be penalized while they're waiting for one PR to be merged or hit production because they're able to post the next PR.

Correct, they cannot. I tried to parallelize them on another issue but had to wait. (Everything clicked there, including C+ staying up till 4 am to match the opposite timezone, the internal engineers being immediately responsive, and me having a commit and everything uploaded before being assigned, so no decrease there, but only just.)
